be/src/runtime/stream_load/stream_load_pipe.cpp

```diff
@@ -275,6 +277,70 @@ Status StreamLoadPipe::_push_front_unlocked(const ByteBufferPtr& buf) {
     return Status::OK();
 }
 
+CompressedStreamLoadPipeReader::CompressedStreamLoadPipeReader(std::shared_ptr<StreamLoadPipe> pipe,
+                                                               TCompressionType::type compression_type)
+        : StreamLoadPipeReader(std::move(pipe)), _compression_type(compression_type) {}
+
+StatusOr<ByteBufferPtr> CompressedStreamLoadPipeReader::read() {
+    size_t buffer_size = DEFAULT_DECOMPRESS_BUFFER_SIZE;
+    if (_decompressor == nullptr) {
+        auto compression = CompressionUtils::to_compression_pb(_compression_type);
+        if (compression == CompressionTypePB::UNKNOWN_COMPRESSION) {
+            return Status::NotSupported("Unsupported compression algorithm: " + std::to_string(_compression_type));
+        }
+        RETURN_IF_ERROR(StreamCompression::create_decompressor(compression, &_decompressor));
+    }
+
+    if (_decompressed_buffer == nullptr) {
+        _decompressed_buffer = ByteBuffer::allocate(buffer_size);
+    }
+
+    ASSIGN_OR_RETURN(auto buf, StreamLoadPipeReader::read());
+
+    // try to read all compressed data into _decompressed_buffer
+    bool stream_end = false;
+    size_t total_bytes_read = 0;
+    size_t bytes_read = 0;
+    size_t bytes_written = 0;
+    RETURN_IF_ERROR(_decompressor->decompress(reinterpret_cast<uint8_t*>(buf->ptr), buf->remaining(), &bytes_read,
+                                              reinterpret_cast<uint8_t*>(_decompressed_buffer->ptr),
+                                              _decompressed_buffer->capacity, &bytes_written, &stream_end));
+    _decompressed_buffer->pos = bytes_written;
+    bytes_written = 0;
+    total_bytes_read += bytes_read;
+
+    std::list<ByteBufferPtr> pieces;
+    size_t pieces_size = 0;
+    // read all pieces
+    while (!stream_end) {
+        // buffer size grows exponentially
+        buffer_size = buffer_size < MAX_DECOMPRESS_BUFFER_SIZE ? buffer_size * 2 : MAX_DECOMPRESS_BUFFER_SIZE;
+        auto piece = ByteBuffer::allocate(buffer_size);
+        RETURN_IF_ERROR(_decompressor->decompress(
+                reinterpret_cast<uint8_t*>(buf->ptr) + total_bytes_read, buf->remaining() - total_bytes_read,
+                &bytes_read, reinterpret_cast<uint8_t*>(piece->ptr), piece->capacity, &bytes_written, &stream_end));
+        piece->pos = bytes_written;
+        pieces.emplace_back(std::move(piece));
+        pieces_size += bytes_written;
+        total_bytes_read += bytes_read;
+        bytes_read = 0;
+        bytes_written = 0;
+    }
+
+    if (pieces_size > 0) {
+        if (_decompressed_buffer->remaining() < pieces_size) {
+            // align to 1024 bytes.
+            auto sz = std::ceil((_decompressed_buffer->pos + pieces_size) * 1.0 / 1024) * 1024;
```

Comment:
Is it possible that sz is 0?

Reply:
Updated. It has been replaced with `ALIGN_UP(xxx, 1024)`